File: frontend/src/types/Goal.ts
export type GoalType = 'EMERGENCY_FUND' | 'PURCHASE' | 'TRAVEL' | 'INVESTMENT' | 'CUSTOM';
export type GoalStatus = 'ACTIVE' | 'COMPLETED' | 'ABANDONED';

export interface Goal {
  id: number;
  name: string;
  description: string;
  goalType: GoalType;
  targetAmount: number;
  currentAmount: number;
  progressPercent: number;
  remaining: number;
  deadline: string | null;
  isOverdue: boolean;
  hpReward: number;
  status: GoalStatus;
  createdAt: string;
  completedAt: string | null;
}

export interface GoalFormData {
  name: string;
  description: string;
  goalType: GoalType;
  targetAmount: number;
  deadline: string;
  hpReward: number;
}

export interface ContributionFormData {
  amount: number;
}

File: backend/src/main/java/com/financetracker/backend/model/Goal.java
package com.financetracker.backend.model;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "goals")
public class Goal {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "goal_type", nullable = false)
    private GoalType goalType;

    @Column(name = "target_amount", nullable = false)
    private Double targetAmount;

    @Column(name = "current_amount", nullable = false)
    private Double currentAmount = 0.0;

    private LocalDate deadline;

    @Column(name = "hp_reward", nullable = false)
    private Integer hpReward = 10;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private GoalStatus status = GoalStatus.ACTIVE;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    // Getters and setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public GoalType getGoalType() { return goalType; }
    public void setGoalType(GoalType goalType) { this.goalType = goalType; }

    public Double getTargetAmount() { return targetAmount; }
    public void setTargetAmount(Double targetAmount) { this.targetAmount = targetAmount; }

    public Double getCurrentAmount() { return currentAmount; }
    public void setCurrentAmount(Double currentAmount) { this.currentAmount = currentAmount; }

    public LocalDate getDeadline() { return deadline; }
    public void setDeadline(LocalDate deadline) { this.deadline = deadline; }

    public Integer getHpReward() { return hpReward; }
    public void setHpReward(Integer hpReward) { this.hpReward = hpReward; }

    public GoalStatus getStatus() { return status; }
    public void setStatus(GoalStatus status) { this.status = status; }

    public LocalDateTime getCreatedAt() { return createdAt; }

    public LocalDateTime getCompletedAt() { return completedAt; }
    public void setCompletedAt(LocalDateTime completedAt) { this.completedAt = completedAt; }
}